statistics/handle/cache/statscacheinner.go

```diff
@@ -140,36 +151,63 @@ func (sc *StatsCache) Version() uint64 {
 
 // Front returns the front element's owner tableID, only used for test.
 func (sc *StatsCache) Front() int64 {
-	return sc.c.Front()
+	return sc.innerCache().Front()
 }
 
-// CopyAndUpdate copies a new cache and updates the new statistics table cache.
-func (sc *StatsCache) CopyAndUpdate(tables []*statistics.Table, deletedIDs []int64, opts ...TableStatsOpt) *StatsCache {
-	option := &TableStatsOption{}
-	for _, opt := range opts {
-		opt(option)
-	}
-	newCache := &StatsCache{c: sc.c.Copy()}
-	newCache.maxTblStatsVer.Store(sc.maxTblStatsVer.Load())
-	for _, tbl := range tables {
-		id := tbl.PhysicalID
-		if option.byQuery {
-			newCache.c.Put(id, tbl, true)
-		} else {
-			newCache.c.Put(id, tbl, false)
+// BatchUpdate updates the StatsCache.
+func (sc *StatsCache) BatchUpdate(tables []*statistics.Table, deletedIDs []int64, opts ...TableStatsOpt) {
+	// current implementation of StatsCacheInner cannot handle concurrent read/write with high performance,
+	// so we update it in a COW manner here to eliminate read/write conflicts.
+	// TODO: remove this COW implementation and support in-place updates.
+	cowUpdate := func() (oldCache, newCache internal.StatsCacheInner) {
+		option := &TableStatsOption{}
+		for _, opt := range opts {
+			opt(option)
+		}
+		oldCache = sc.innerCache()
+		newCache = oldCache.Copy()
+		for _, tbl := range tables {
+			id := tbl.PhysicalID
+			if option.byQuery {
+				newCache.Put(id, tbl, true)
+			} else {
+				newCache.Put(id, tbl, false)
+			}
 		}
+		for _, id := range deletedIDs {
+			newCache.Del(id)
+		}
+		return oldCache, newCache
 	}
-	for _, id := range deletedIDs {
-		newCache.c.Del(id)
+
+	for {
+		oldCache, newCache := cowUpdate()
+		if sc.c.CompareAndSwap(&oldCache, &newCache) {
+			break
+		} // sc.c has been updated by some other thread, so we need to update it again.
 	}
 
 	// update the maxTblStatsVer
+	var maxVer uint64
 	for _, t := range tables {
-		if t.Version > newCache.maxTblStatsVer.Load() {
-			newCache.maxTblStatsVer.Store(t.Version)
+		if maxVer < t.Version {
+			maxVer = t.Version
 		}
 	}
-	return newCache
+	for v := sc.maxTblStatsVer.Load(); v < maxVer; v = sc.maxTblStatsVer.Load() {
+		if sc.maxTblStatsVer.CompareAndSwap(v, maxVer) {
+			break
+		} // other goroutines have updated the sc.maxTblStatsVer, so we need to check again.
```

Comment:
It looks kind of strange that CAS operations of cache pointer and version are separated.